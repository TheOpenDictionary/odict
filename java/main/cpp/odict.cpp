#include <jni.h>
#include <stdio.h>
#include <iostream>
#include "bridge/bridge.h"

using namespace std;

#ifndef CGO_EXPORT_BRIDGE_EXISTS
#error bridging header not found
#endif

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_lookup(JNIEnv *env, jobject, jstring query, jstring dictID)
{
  const char *dictionaryID = env->GetStringUTFChars(dictID, 0);
  const char *entry_term = env->GetStringUTFChars(query, 0);
  char *result = LookupEntry((char *)entry_term, (char *)dictionaryID);
  jstring str = env->NewStringUTF(result);

  free(result);

  return str;
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_index(JNIEnv *env, jobject, jstring path, jboolean force)
{
  char *p = (char *)env->GetStringUTFChars(path, 0);
  
  IndexDictionary(p, force);

  env->ReleaseStringUTFChars(path, p);
}

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_search(JNIEnv *env, jobject, jstring query, jstring dictionary_id)
{
  char *q = (char *)env->GetStringUTFChars(query, 0);
  char *dictid = (char *)env->GetStringUTFChars(dictionary_id, 0);
  const char *result = SearchDictionary(q, dictid);

  jstring res = env->NewStringUTF(result);

  env->ReleaseStringUTFChars(query, q);
  env->ReleaseStringUTFChars(dictionary_id, dictid);

  free((char *)result);

  return res;
}

extern "C" JNIEXPORT jstring JNICALL
Java_org_odict_Dictionary_read(JNIEnv *env, jobject, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);
  char *encoded = ReadDictionary((char *)dictionary_path);

  jstring res = env->NewStringUTF(encoded);

  env->ReleaseStringUTFChars(path, dictionary_path);

  free(encoded);

  return res;
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_write(JNIEnv *env, jobject, jstring xml, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);
  const char *content = env->GetStringUTFChars(xml, 0);

  WriteDictionary((char *)content, (char *)dictionary_path);

  env->ReleaseStringUTFChars(path, dictionary_path);
  env->ReleaseStringUTFChars(xml, content);
}

extern "C" JNIEXPORT void JNICALL
Java_org_odict_Dictionary_compile(JNIEnv *env, jobject, jstring path)
{
  const char *dictionary_path = env->GetStringUTFChars(path, 0);

  CompileDictionary((char *)dictionary_path);

  env->ReleaseStringUTFChars(path, dictionary_path);
}
