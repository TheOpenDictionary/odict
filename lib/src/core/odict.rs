/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use rkyv::access_unchecked;

use crate::{semver::SemanticVersion, ArchivedDictionary, Dictionary};

#[derive(Clone)]
pub struct OpenDictionary {
    pub signature: String,
    pub version: SemanticVersion,
    pub path: Option<PathBuf>,
    pub total_size: u64,

    pub(super) content: Vec<u8>,
}

impl OpenDictionary {
    pub fn access(&self) -> crate::Result<&ArchivedDictionary> {
        let archived = unsafe { access_unchecked::<crate::ArchivedDictionary>(&self.content[..]) };
        Ok(archived)
    }

    pub fn deserialize(&self) -> crate::Result<Dictionary> {
        let dict: Dictionary = self.access()?.deserialize()?;
        Ok(dict)
    }
}
