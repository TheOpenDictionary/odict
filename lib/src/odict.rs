/* -------------------------------------------------------------------------- */
/*                               DictionaryFile                               */
/* -------------------------------------------------------------------------- */

use std::path::PathBuf;

use crate::version::SemanticVersion;

#[derive(Clone)]
pub struct OpenDictionary {
    pub(crate) signature: String,
    pub(crate) version: SemanticVersion,
    pub(crate) path: Option<PathBuf>,
    pub(crate) bytes: Vec<u8>,
}

impl OpenDictionary {
    pub fn signature(&self) -> &str {
        &self.signature
    }

    pub fn path(&self) -> &Option<PathBuf> {
        &self.path
    }

    pub fn version(&self) -> &SemanticVersion {
        &self.version
    }

    pub fn size(&self) -> crate::Result<u64> {
        if let Some(path) = &self.path {
            if let Ok(metadata) = std::fs::metadata(path) {
                return Ok(metadata.len());
            }
        }

        let bytes = self.to_bytes()?;

        Ok(bytes.len().try_into()?)
    }

    pub fn contents(&self) -> crate::Result<&crate::schema::ArchivedDictionary> {
        Ok(rkyv::access::<
            crate::schema::ArchivedDictionary,
            rkyv::rancor::Error,
        >(&self.bytes[..])?)
    }
}
