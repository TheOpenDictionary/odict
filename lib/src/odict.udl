namespace odict {
};

interface DictionaryFile {};

interface DictionaryReader {
  constructor();
};