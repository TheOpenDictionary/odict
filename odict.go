package main

import (
	"log"
	"os"

	"github.com/TheOpenDictionary/odict/cli"
	"github.com/TheOpenDictionary/odict/lib/validator"
)

func main() {
	validator.InitializeValidator()

	defer validator.CleanupValidator()

	err := cli.App.Run(os.Args)

	if err != nil {
		log.Fatal(err)
	}
}
