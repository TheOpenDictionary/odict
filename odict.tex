\documentclass{article}
\usepackage{hyperref}
\begin{document}
\title{\textbf{The Open Dictionary Project}}
\author{\textbf{Tyler Nickerson} \\ \small{Linguistic Inc.} \\ \small{tyler@linguistic.io}}
\date{}
\maketitle
The current state of digital dictionary technology presents numerous challenges for those seeking offline and programmatic access to lexical data. Most dictionaries still in use today are built upon outdated or proprietary technologies, making them challenging to adopt in modern applications. This lack of standardization has led to a proliferation of custom protocols for storing data across many offline dictionary applications. This problem is further compounded by the fact that lexemes are typically stored as HTML fragments, making the extraction of structured data both difficult and variable across different formats.

In response to these challenges, we propose the Open Dictionary Project (ODict) as a solution for creating a new, open standard for dictionary formats that is both extensible and user-friendly. ODict includes a public file format specification, a command-line compiler, as well as wrapper libraries for creating and accessing lexical dictionaries. Dictionaries are compiled from human-readable XML and stored in a single, compressed binary file that can be easily transmitted and accessed on any device. Lexical information, such as etymologies, senses, and definitions, is stored in a buffer as structured data that can be output in a variety of formats. This approach allows language scientists to leverage well-formed lexical data in their own projects without having to process HTML.

Lexemes can be instantly retrieved \emph{without} deserializing the file buffer and \emph{without} relying on external index files, making lookups extremely fast. Furthermore, unlike other formats, ODict features official language bindings for NodeJS, Go, Python, and JVM languages (although the ODict specification can be used in any programming language supported by its underlying FlatBuffer protocol).

In this presentation, the process for writing, compiling, and consuming ODict files will be demonstrated, along with a discussion around the motivation and development of the format as described above. We will also share our future plans for the Open Dictionary Project in order to encourage a constructive discourse among conference attendees. The development of a new, open standard for dictionary formats is an important step towards improving the accessibility and reliability of lexical data for all users. We believe ODict has great potential in the fields of natural language processing and language learning, and we hope to see the format adopted by the wider community.
\end{document}
