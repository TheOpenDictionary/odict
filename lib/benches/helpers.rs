use std::sync::LazyLock;

use odict::{DictionaryFile, DictionaryReader, DictionaryWriter, Result};

pub fn create_archive_dict(name: &str) -> Result<DictionaryFile> {
    let reader = DictionaryReader::default();
    let writer = DictionaryWriter::default();
    let input = format!("../examples/{}.xml", name);
    let output = format!("../examples/{}.odict", name);

    writer.compile_xml(&input, &output)?;
    reader.read_from_path(&output)
}

pub static EXAMPLE_DICTIONARY_1: LazyLock<DictionaryFile> =
    LazyLock::new(|| create_archive_dict("wiktionary").unwrap());
