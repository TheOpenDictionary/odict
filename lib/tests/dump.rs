mod helpers;

mod dump_tests {

    use insta::assert_snapshot;

    use odict::{json::ToJSON, xml::ToXML, Dictionary, ID};

    use crate::helpers::EXAMPLE_DICT_1;

    #[test]
    fn test_xml() {
        let dict = EXAMPLE_DICT_1.to_dictionary().unwrap();

        let dumped = dict.clone().to_xml(true).unwrap();

        let mut dictionary2 = Dictionary::from(dumped.as_str()).unwrap();

        dictionary2.id = dict.id.clone();

        assert_eq!(dict, dictionary2);
    }

    #[test]
    fn test_json() {
        let mut mutable = EXAMPLE_DICT_1.to_dictionary().unwrap().clone();

        // Keep it consistent
        mutable.id = ID::parse("2ee2a1ae-f7ff-4590-ba2d-de857ba7857f").unwrap();

        let dumped = mutable.to_json(true).unwrap();

        assert_snapshot!(dumped);
    }
}
