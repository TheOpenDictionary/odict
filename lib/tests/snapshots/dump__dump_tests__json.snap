---
source: lib/tests/dump.rs
expression: dumped
snapshot_kind: text
---
{
  "id": "2ee2a1ae-f7ff-4590-ba2d-de857ba7857f",
  "name": "Example Dictionary 1",
  "entries": {
    "cat": {
      "term": "cat",
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "pos": "n",
              "definitions": [
                {
                  "type": "definition",
                  "value": "a cat",
                  "examples": [
                    "There goes a cat!"
                  ],
                  "notes": [
                    {
                      "value": "Some definition note",
                      "examples": [
                        {
                          "$text": "Some example"
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          }
        }
      ]
    },
    "dog": {
      "term": "dog",
      "etymologies": [
        {
          "pronunciation": "dooooog",
          "description": "Latin root",
          "senses": {
            "un": {
              "pos": "un",
              "definitions": [
                {
                  "type": "definition",
                  "value": "a dog"
                }
              ]
            }
          }
        }
      ]
    },
    "poo": {
      "term": "poo",
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "un": {
              "pos": "un",
              "definitions": [
                {
                  "type": "group",
                  "description": "A number of verb senses",
                  "definitions": [
                    {
                      "value": "crap, shit"
                    }
                  ]
                }
              ]
            }
          }
        }
      ]
    },
    "ran": {
      "term": "ran",
      "see_also": "run"
    },
    "run": {
      "term": "run",
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "pos": "n",
              "definitions": [
                {
                  "type": "definition",
                  "value": "(vertebrates) To move swiftly",
                  "examples": [
                    "The dog runs after the cat.",
                    "The horse ran away."
                  ]
                },
                {
                  "type": "definition",
                  "value": "Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip."
                },
                {
                  "type": "definition",
                  "value": "A pleasure trip."
                },
                {
                  "type": "definition",
                  "value": "Flight, instance or period of fleeing."
                },
                {
                  "type": "definition",
                  "value": "Migration (of fish)."
                },
                {
                  "type": "definition",
                  "value": "A group of fish that migrate, or ascend a river for the purpose of spawning."
                }
              ]
            },
            "v": {
              "pos": "v",
              "definitions": [
                {
                  "type": "group",
                  "description": "A number of verb senses",
                  "definitions": [
                    {
                      "value": "(vertebrates) To move swiftly (figuratively)",
                      "examples": [
                        "The dog runs after the cat.",
                        "The horse ran away."
                      ]
                    },
                    {
                      "value": "(fluids) To flow."
                    },
                    {
                      "value": "(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled."
                    },
                    {
                      "value": "(social) To carry out an activity."
                    },
                    {
                      "value": "To extend or persist, statically or dynamically, through space or time."
                    },
                    {
                      "value": "(transitive) To execute or carry out a plan, procedure or program."
                    }
                  ]
                }
              ]
            }
          }
        }
      ]
    }
  }
}
